pkg/sac/allow_fixed_scopes_checker_core.go

```diff
@@ -3,121 +3,256 @@ package sac
+func allowFixedAccessModeLevelScopes(
+	accessLevelKeys []ScopeKey,
+) ScopeCheckerCore {
+	return &allowedFixedScopesCheckerCore{
+		checkerLevel:  GlobalScopeKind,
+		accessKeys:    typedKeySet[AccessModeScopeKey](accessLevelKeys),
+		resourceKeys:  typedKeySet[ResourceScopeKey]([]ScopeKey{}),
+		clusterKeys:   typedKeySet[ClusterScopeKey]([]ScopeKey{}),
+		namespaceKeys: typedKeySet[NamespaceScopeKey]([]ScopeKey{}),
 	}
-	if _, ok := c.topLevelKeys()[key]; ok {
-		return c.next()
+}
+
+func allowFixedResourceLevelScopes(
+	accessLevelKeys []ScopeKey,
+	resourceLevelKeys []ScopeKey,
+) ScopeCheckerCore {
+	return &allowedFixedScopesCheckerCore{
+		checkerLevel:  GlobalScopeKind,
+		accessKeys:    typedKeySet[AccessModeScopeKey](accessLevelKeys),
+		resourceKeys:  typedKeySet[ResourceScopeKey](resourceLevelKeys),
+		clusterKeys:   typedKeySet[ClusterScopeKey]([]ScopeKey{}),
+		namespaceKeys: typedKeySet[NamespaceScopeKey]([]ScopeKey{}),
 	}
-	return denyAllScopeCheckerCore
 }
 
-func (c allowFixedScopesCheckerCore) EffectiveAccessScope(resource permissions.ResourceWithAccess) (*effectiveaccessscope.ScopeTree, error) {
-	if len(c) == 0 {
-		return effectiveaccessscope.UnrestrictedEffectiveAccessScope(), nil
+func allowFixedClusterLevelScopes(
+	accessLevelKeys []ScopeKey,
+	resourceLevelKeys []ScopeKey,
+	clusterLevelKeys []ScopeKey,
+) ScopeCheckerCore {
+	return &allowedFixedScopesCheckerCore{
+		checkerLevel:  GlobalScopeKind,
+		accessKeys:    typedKeySet[AccessModeScopeKey](accessLevelKeys),
+		resourceKeys:  typedKeySet[ResourceScopeKey](resourceLevelKeys),
+		clusterKeys:   typedKeySet[ClusterScopeKey](clusterLevelKeys),
+		namespaceKeys: typedKeySet[NamespaceScopeKey]([]ScopeKey{}),
 	}
-	// TODO (ROX-9981): Change the structure of allowFixedScopesCheckerCore to have explicit scope level semantic
-	for key := range c.topLevelKeys() {
-		switch key.(type) {
-		case AccessModeScopeKey:
-			return c.getAccessModeEffectiveAccessScope(resource)
-		case ResourceScopeKey:
-			return c.getResourceEffectiveAccessScope(resource)
-		case ClusterScopeKey:
-			return c.getClusterEffectiveAccessScope()
+}
+
+func allowFixedNamespaceLevelScopes(
+	accessLevelKeys []ScopeKey,
+	resourceLevelKeys []ScopeKey,
+	clusterLevelKeys []ScopeKey,
+	namespaceLevelKeys []ScopeKey,
+) ScopeCheckerCore {
+	return &allowedFixedScopesCheckerCore{
+		checkerLevel:  GlobalScopeKind,
+		accessKeys:    typedKeySet[AccessModeScopeKey](accessLevelKeys),
+		resourceKeys:  typedKeySet[ResourceScopeKey](resourceLevelKeys),
+		clusterKeys:   typedKeySet[ClusterScopeKey](clusterLevelKeys),
+		namespaceKeys: typedKeySet[NamespaceScopeKey](namespaceLevelKeys),
+	}
+}
+
+func (c *allowedFixedScopesCheckerCore) SubScopeChecker(scopeKey ScopeKey) ScopeCheckerCore {
+	switch key := scopeKey.(type) {
+	case AccessModeScopeKey:
+		if c.checkerLevel != GlobalScopeKind {
+			return denyAllScopeCheckerCore
+		}
+		if c.accessKeys.Cardinality() > 0 && !c.accessKeys.Contains(key) {
```

Comment:
do we need cardinality check? If empty then it will not contain any keys.
```suggestion
		if !c.accessKeys.Contains(key) {
```

Reply:
The previous logic in drill-down was:
- if there is no next-levels key set, return a full-access scope checker core
- if there is a next-level key set, check if the requested drill-down key is in the next level set. If so, drill down, otherwise return denyAll

The re-implemented logic is slightly flawed. I'll fix that.